src/sync.js

```diff
@@ -0,0 +1,128 @@
+      'in the `syncHistoryWithStore()` options. ' +
+      'Ensure you have added the `routerReducer` to your store\'s ' +
+      'reducers via `combineReducers` or whatever method you use to isolate ' +
+      'your reducers.'
+    )
+  }
+
+  let initialLocation
+  let currentLocation
+  let isTimeTraveling
+  let unsubscribeFromStore
+  let unsubscribeFromHistory
+
+  // What does the store say about current location?
+  const getLocationInStore = (useInitialIfEmpty) => {
+    const locationState = selectLocationState(store.getState())
+    return locationState.locationBeforeTransitions ||
+      (useInitialIfEmpty ? initialLocation : undefined)
+  }
+
+  // If the store is replayed, update the URL in the browser to match.
+  if (adjustUrlOnReplay) {
+    const handleStoreChange = () => {
+      const locationInStore = getLocationInStore(true)
+      if (currentLocation === locationInStore) {
+        return
+      }
+
+      // Update address bar to reflect store state
+      isTimeTraveling = true
+      currentLocation = locationInStore
+      history.transitionTo(Object.assign({},
+        locationInStore,
+        { action: 'PUSH' }
+      ))
+      isTimeTraveling = false
+    }
+
+    unsubscribeFromStore = store.subscribe(handleStoreChange)
+    handleStoreChange()
+  }
+
+  // Whenever location changes, dispatch an action to get it in the store
+  const handleLocationChange = (location) => {
+    // ... unless we just caused that location change
+    if (isTimeTraveling) {
+      return
+    }
+
+    // Remember where we are
+    currentLocation = location
+
+    // Are we being called for the first time?
+    if (!initialLocation) {
+      // Remember as a fallback in case state is reset
+      initialLocation = location
+
+      // Respect persisted location, if any
+      if (getLocationInStore()) {
+        return
+      }
+    }
+
+    // Tell the store to update by dispatching an action
+    store.dispatch({
+      type: LOCATION_CHANGE,
+      locationBeforeTransitions: location
+    })
+  }
+  unsubscribeFromHistory = history.listen(handleLocationChange)
+
+  // The enhanced history uses store as source of truth
+  return Object.assign({}, history, {
+    // The listeners are subscribed to the store instead of history
+    listen(listener) {
+      // History listeners expect a synchronous call
+      listener(getLocationInStore(true))
+
+      // Keep track of whether we unsubscribed, as Redux store
+      // only applies changes in subscriptions on next dispatch
+      let unsubscribed = false
+      const unsubscribeFromStore = store.subscribe(() => {
+        if (!unsubscribed) {
+          listener(getLocationInStore(true))
+        }
+      })
+
+      // Let user unsubscribe later
+      return () => {
+        unsubscribed = true
+        unsubscribeFromStore()
+      }
+    },
+
+    // It also provides a way to destroy internal listeners
+    unsubscribe() {
```

Comment:
How does disposing history enhancers compose? Is this the only history enhancer with internal resources?
If normally enhancers free resources when last subscription dies, can we do this here? If not, why?
If we decide unsubscribe() is way to go, should we not call it on underlying history too, if it exists, to ensure this enhancer is composable with other possible enhancers that may also have internal resources?

In the original PR I called this dispose() to be consistent with disposables in Rx and make it clear that this method is not related to listen() subscriptions. And that using enhanced history() after dispose() won't work. Essentially we might want to throw it you attempt to listen() to a disposed history to prevent silly errors. What do you think?